File: Programowanie_Obiektowe/PO_C/PO_C/Barszcz_Sosnowskiego.cpp
#include "Barszcz_Sosnowskiego.h"

Barszcz_Sosnowskiego::Barszcz_Sosnowskiego(int x, int y, Swiat& swiat) : Roslina(BARSZCZ_SOSNOWSKIEGO_SILA, x, y, swiat) {}

void Barszcz_Sosnowskiego::akcja()
{
    wiek++;

    if (rand() % 100 < PRAWDOPODOBIENSTWO_ROZPRZESTRZENIANIA)
    {
        rozmnazanie();
    }
    int dx[8] = { -1, 0, 1, -1, 1, -1, 0, 1 };
    int dy[8] = { 1, 1, 1, 0, 0, -1, -1, -1 };

    for (int i = 0; i < 8; i++)
    {
        int newX = x + dx[i];
        int newY = y + dy[i];

        if (swiat.czyNaPlanszy(newX, newY))
        {
            Organizm* cel = swiat.getOrganizmNa(newX, newY);
            if (cel != nullptr && cel->getTyp() == ZWIERZE && cel->rysowanie() != CYBER_OWCA)         //barszcz zabija zwierze
            {
                swiat.setLog(swiat.getLog() + string(1, rysowanie()) + " (" + to_string(x) + ", " + to_string(y) + ") zabija " + string(1, cel->rysowanie()) + " (" + to_string(cel->getX()) + ", " + to_string(cel->getY()) + ")" + "\n");
                swiat.ustawOrganizmNa(cel->getY(), cel->getX(), nullptr);
                swiat.usunOrganizm(cel);
            }
        }
    }
}

void Barszcz_Sosnowskiego::kolizja(Organizm& atakujacy)
{
    swiat.setLog(swiat.getLog() + string(1, rysowanie()) + " (" + to_string(x) + ", " + to_string(y) + ") zjedzony przez " + string(1, atakujacy.rysowanie()) + " (" + to_string(atakujacy.getX()) + ", " + to_string(atakujacy.getY()) + ")" + "\n");
    if (atakujacy.rysowanie() != CYBER_OWCA)
    {
        swiat.setLog(swiat.getLog() + string(1, atakujacy.rysowanie()) + " (" + to_string(atakujacy.getX()) + ", " + to_string(atakujacy.getY()) + ") zatruty przez " + string(1, rysowanie()) + "\n");
        swiat.ustawOrganizmNa(atakujacy.getY(), atakujacy.getX(), nullptr);
        swiat.usunOrganizm(&atakujacy);
    }
    else
    {
        swiat.setLog(swiat.getLog() + string(1, atakujacy.rysowanie()) + " (" + to_string(atakujacy.getX()) + ", " + to_string(atakujacy.getY()) + ") strawil " + string(1, rysowanie()) + "\n");
    }
    swiat.ustawOrganizmNa(y, x, nullptr);
    swiat.usunOrganizm(this);
}

char Barszcz_Sosnowskiego::rysowanie() const
{
    return BARSZCZ_SOSNOWSKIEGO;
}

Organizm* Barszcz_Sosnowskiego::stworzPotomka(int x, int y) const
{
    return new Barszcz_Sosnowskiego(x, y, swiat);
}

File: Programowanie_Obiektowe/PO_JAVA/src/Barszcz_Sosnowskiego.java
import java.util.List;
import java.util.Random;

public class Barszcz_Sosnowskiego extends Roslina
{
    public Barszcz_Sosnowskiego(int x, int y, Swiat swiat)
    {
        super(Stale.BARSZCZ_SOSNOWSKIEGO_SILA, x, y, swiat);
    }

    @Override
    public void akcja()
    {
        wiek++;

        if (new Random().nextInt(100) < Stale.PRAWDOPODOBIENSTWO_ROZPRZESTRZENIANIA)
        {
            rozmnazanie();
        }

        List<int[]> sasiedzi = swiat.sasiedniePola(x, y);

        for (int[] pole : sasiedzi)
        {
            int newX = pole[0];
            int newY = pole[1];

            Organizm cel = swiat.getOrganizmNa(newX, newY);
            if (cel != null && cel.getTyp() == Stale.TypOrganizmu.ZWIERZE && cel.rysowanie() != Stale.CYBER_OWCA)
            {
                swiat.dodajLog(rysowanie() + " (" + x + ", " + y + ") zabija " + cel.rysowanie() + " (" + cel.getX() + ", " + cel.getY() + ")");

                swiat.ustawOrganizmNa(cel.getX(), cel.getY(), null);
                swiat.usunOrganizm(cel);
            }
        }
    }

    @Override
    public void kolizja(Organizm atakujacy)
    {
        swiat.dodajLog(rysowanie() + " (" + x + ", " + y + ") zjedzony przez " + atakujacy.rysowanie() + " (" + atakujacy.getX() + ", " + atakujacy.getY() + ")");

        if (atakujacy.rysowanie() != Stale.CYBER_OWCA)
        {
            swiat.dodajLog(atakujacy.rysowanie() + " (" + atakujacy.getX() + ", " + atakujacy.getY() + ") zatruty przez " + rysowanie());

            swiat.ustawOrganizmNa(atakujacy.getX(), atakujacy.getY(), null);
            swiat.usunOrganizm(atakujacy);
        }
        else
        {
            swiat.dodajLog(atakujacy.rysowanie() + " (" + atakujacy.getX() + ", " + atakujacy.getY() + ") strawił " + rysowanie());
        }

        swiat.ustawOrganizmNa(x, y, null);
        swiat.usunOrganizm(this);
    }

    @Override
    public char rysowanie()
    {
        return Stale.BARSZCZ_SOSNOWSKIEGO;
    }

    @Override
    public String nazwa()
    {
        return Stale.BARSZCZ_SOSNOWSKIEGO_S;
    }

    @Override
    public Organizm stworzPotomka(int x, int y)
    {
        return new Barszcz_Sosnowskiego(x, y, swiat);
    }
}
